internal/cli/cmd/apply/apply.go

```diff
@@ -19,8 +19,54 @@ import (
 	kargosvcapi "github.com/akuity/kargo/pkg/api/service/v1alpha1"
 )
 
+type applyOptions struct {
+	*option.Option
+
+	Filenames []string
+}
+
+// newApplyOptions creates a new options struct.
+func newApplyOptions(opt *option.Option) *applyOptions {
+	return &applyOptions{Option: opt}
+}
+
+// AddFlags adds the flags for the apply options to the provided command.
+func (o *applyOptions) AddFlags(cmd *cobra.Command) {
```

Comment:
I'm not dead set against this, but I am curious about the rationale for extracting this logic from `NewCommand()` into its own function.

I ask because I think the code is _slightly_ easier to follow the closer we keep all the flag setup to the actual definition of the command... _but_ I suspect you had reasons for this.

Reply:
Nit: It could be unexported.

Reply:
The goal is to move into a direction in which the actual logic of running the command is driven by just the options struct, making it easier to test things in isolation (without requiring a full end-to-end setup).

The construction of the command then just becomes gluing various building blocks together, in which case I think a clear separation of concerns becomes desirable.

I also think it helps "drive by contributors" to know where to put what, while providing guardrails to keep the code structured and tidy. If they want to add some functionality based on a value, they now have clear steps to follow:

1. Add the field to the struct.
2. Register the flag in the designated area.
3. Implement logic which consumes the value.

Lastly, it makes options "reusable". In case you have hypothetical options which are e.g. a subset of other options for multiple commands. You can simply embed the options struct and register the flags in multiple commands simultaneously.

Reply:
> The goal is to move into a direction in which the actual logic of running the command is driven by just the options struct, making it easier to test things in isolation (without requiring a full end-to-end setup).

Agree and wish we'd done things this from the start.

> gluing various building blocks together

Agree... with the hope of keeping the glue easy to follow, which was why I raised this question. My only worry had been that breaking flag registration out into its own function undermined keeping the glue in one place. In my head, something like this might tick all your boxes and mine:

```golang
func NewCommand(cfg config.CLIConfig, opt *option.Option) *cobra.Command {
	cmdOpts := &applyOptions{Option: opt}

	cmd := &cobra.Command{
		Use:   "apply [--project=project] -f (FILENAME)",
		Short: "Apply a resource from a file or from stdin",
		PreRunE: func(cmd *cobra.Command, _ []string) error {
			return cmdOpts.Validate()
		},
		RunE: func(cmd *cobra.Command, _ []string) error {
			// Run should be easy to test without involving Cobra.
			return run(cmd.Context(), cmdOpts)
		},
	}

	opt.PrintFlags.AddFlags(cmd)
	// etc...
	if err := cmd.MarkFlagDirname(option.FilenameFlag); err != nil {
		panic(errors.Wrap(err, "could not mark filename flag as dirname"))
	}

	return cmd
}
```

You can look at this one `NewCommand` func now for quite a concise overview of the whole command. Admittedly, having the logic in-line as we previously did stood in the way of that even more than breaking flag registration into its own function ever did. Extracting that logic makes things look less insane.

wdyt?

Reply:
Not strongly opposed to it, although I think that with a bigger number of flags and/or markers you can end up in the same position as the in-line run. An example:

```go
func NewCommand(cfg config.CLIConfig, opt *option.Option) *cobra.Command {
	cmdOpts := newPromotionOptions(opt)

	cmd := &cobra.Command{
		Use:   "promote [--project=project] --freight=freight-id [--stage=stage] [--subscribers-of=stage]",
		Short: "Manage the promotion of freight",
		Args:  option.NoArgs,
		Example: `
# Promote a freight to a stage for a specific project
kargo promote --project=my-project --freight=abc123 --stage=dev

# Promote a freight to subscribers of a stage for a specific project
kargo promote --project=my-project --freight=abc123 --subscribers-of=dev

# Promote a freight to a stage for the default project
kargo config set project my-project
kargo promote --freight=abc123 --stage=dev

# Promote a freight to subscribers of a stage for the default project
kargo config set project my-project
kargo promote --freight=abc123 --subscribers-of=dev
`,
		PreRunE: func(cmd *cobra.Command, args []string) error {
			return cmdOpts.Validate()
		},
		RunE: func(cmd *cobra.Command, args []string) error {
			return run(cmd.Context(), cmdOpts)
		},
	}

	// This is where the messier part starts.
	o.PrintFlags.AddFlags(cmd)

	option.InsecureTLS(cmd.PersistentFlags(), o.Option)
	option.LocalServer(cmd.PersistentFlags(), o.Option)

	option.Project(cmd.Flags(), &o.Project, o.Project,
		"The Project the Freight belongs to. If not set, the default project will be used.")
	option.Freight(cmd.Flags(), &o.Freight, "The ID of the Freight to promote.")
	option.Stage(cmd.Flags(), &o.Stage, fmt.Sprintf("The Stage to promote the Freight to. If set, --%s "+
		"must not be set.", option.SubscribersOfFlag))
	option.SubscribersOf(cmd.Flags(), &o.SubscribersOf, fmt.Sprintf("The Stage from which the subscribers "+
		"will be used to promote the Freight to. If set, --%s must not be set.", option.StageFlag))

	if err := cmd.MarkFlagRequired(option.FreightFlag); err != nil {
		panic(errors.Wrap(err, "could not mark freight flag as required"))
	}
	cmd.MarkFlagsOneRequired(option.StageFlag, option.SubscribersOfFlag)
	cmd.MarkFlagsMutuallyExclusive(option.StageFlag, option.SubscribersOfFlag)

	return cmd
```

To me, the above looks like something which I still may need to break up into smaller pieces to improve readability, which I think the `AddFlags` method does in the right way by encapsulating the registration with the options themselves (just like `Validate`).

But as said, I do not have extremely strong feelings about this and would be happy to remove and/or circle back to it once we both observe of it becoming a problem.

Reply:
I see your point. There may be no perfect solution here and I don't want to let perfect be the enemy of good. I am also not strongly attached to my opinion. I think stick with what you originally proposed and we'll just keep an eye out for opportunities to improve this further as this evolves.

Reply:
Putting some closure on this. We spoke offline and explored some tweaks to the pattern that look like they're just what we're looking for.